tests/helpers/kube.go

```diff
@@ -610,3 +648,703 @@ func CreateRebootNodeCR(
+	}
+}
+
+// ExecInPod executes a command in a pod and returns stdout, stderr
+func ExecInPod(ctx context.Context, restConfig *rest.Config, namespace, podName, containerName string, command []string) (string, string, error) {
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create clientset: %w", err)
+	}
+
+	req := clientset.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(podName).
+		Namespace(namespace).
+		SubResource("exec").
+		VersionedParams(&v1.PodExecOptions{
+			Container: containerName,
+			Command:   command,
+			Stdin:     false,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(restConfig, "POST", req.URL())
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create executor: %w", err)
+	}
+
+	var stdout, stderr bytes.Buffer
+	err = exec.StreamWithContext(ctx, remotecommand.StreamOptions{
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+
+	return stdout.String(), stderr.String(), err
+}
+
+func IsPodReady(pod v1.Pod) bool {
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == v1.PodReady {
+			return condition.Status == v1.ConditionTrue
+		}
+	}
+	return false
+}
+
+// PortForwardPod sets up port forwarding to a pod and returns channels to control it.
+// Returns (stopChan, readyChan) where:
+// - stopChan: close this to stop port forwarding
+// - readyChan: will be closed when port-forward is ready
+func PortForwardPod(ctx context.Context, restConfig *rest.Config, namespace, podName string, localPort, podPort int) (chan struct{}, chan struct{}) {
+	stopChan := make(chan struct{}, 1)
+	readyChan := make(chan struct{})
+
+	go func() {
+		clientset, err := kubernetes.NewForConfig(restConfig)
+		if err != nil {
+			close(readyChan)
+			return
+		}
+
+		req := clientset.CoreV1().RESTClient().Post().
+			Resource("pods").
+			Namespace(namespace).
+			Name(podName).
+			SubResource("portforward")
+
+		transport, upgrader, err := spdy.RoundTripperFor(restConfig)
+		if err != nil {
+			close(readyChan)
+			return
+		}
+
+		dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+
+		ports := []string{fmt.Sprintf("%d:%d", localPort, podPort)}
+
+		devNull := &bytes.Buffer{}
+
+		fw, err := portforward.New(dialer, ports, stopChan, readyChan, devNull, devNull)
+		if err != nil {
+			close(readyChan)
+			return
+		}
+
+		if err := fw.ForwardPorts(); err != nil {
+			slog.Error("Error forwarding ports", "error", err)
+			return
+		}
+	}()
+
+	return stopChan, readyChan
+}
+
+// ApplyKWOKStage applies a KWOK Stage resource from a YAML file.
+// KWOK Stages are used to customize the behavior of fake nodes and pods.
+// Reference: https://github.com/kubernetes-sigs/kwok/discussions/926
+func ApplyKWOKStage(ctx context.Context, t *testing.T, client klient.Client, filePath string) error {
```

Comment:
Can the kwok state be applied at startup, i.e. in the tilt file? Does it have to be done dynamically?